
## Import packages
```{r}
library(rGREAT)
library(genomation)
library(methylKit)
library(tidyverse)
library(ggbio)
library(GenomicFeatures)
library(GenomicRanges)
library(ggplot2)
library(gridExtra)

output_dir <- "/data/lvisser/post_analysis"
```

```{r}
DE_results_human_Group_R_S_MM <- read.csv("/data/lvisser/DE_genes/DE_results_human_Group_R_S_MM.csv") %>%
  filter(padj < 0.05)
DE_results_human_Group_R_S_NB <- read.csv("/data/lvisser/DE_genes/DE_results_human_Group_R_S_NB.csv") %>%
  filter(padj < 0.05)
DE_results_human_Group_R_Ss_sensSHY <- read.csv("/data/lvisser/DE_genes/DE_results_human_Group_R_Ss_sensSHY.csv") %>%
  filter(padj < 0.05)

# Create a list of dataframes
DE_results_list <- list(
  "R_S_MM" = DE_results_human_Group_R_S_MM,
  "R_S_NB" = DE_results_human_Group_R_S_NB,
  "R_Ss_sensSHY" = DE_results_human_Group_R_Ss_sensSHY
)
```

```{r}
gene.obj <- readTranscriptFeatures("/data/lvisser/segmentations/genes_refseq_all_ucsc.bed")
head(gene.obj)
```

```{r}
cgi.obj <- readFeatureFlank("/data/lvisser/segmentations/cpg_islands_ucsc_cleaned.bed",
                           feature.flank.name=c("CpGi","shores"))
head(cgi.obj)
```

```{r}
reg.obj <- readGeneric("/data/lvisser/segmentations/encode_ccres_ucsc.bed", keep.all.metadata = TRUE, header = TRUE)
reg.obj <- split(reg.obj, mcols(reg.obj)$ucscLabel)
head(reg.obj)
```

```{r}
head(reg.obj)
```

```{r}
annotate_regions <- function(df, output_dir, name, gene.obj, cgi.obj, reg.obj) {
  # Separate hypermethylated and hypomethylated regions
  df_hyper <- df[df$direction == "M",]
  df_hypo <- df[df$direction == "U",]

  # Calculate percentages
  percent_hyper <- round(nrow(df_hyper) / nrow(df) * 100, 2)
  percent_hypo <- round(nrow(df_hypo) / nrow(df) * 100, 2)

  # Annotate regions for both hypermethylated and hypomethylated
  geneDMRann_hyper <- annotateWithGeneParts(as(df_hyper,"GRanges"),gene.obj)
  geneDMRann_hypo <- annotateWithGeneParts(as(df_hypo,"GRanges"),gene.obj)

  cgiDMRann_hyper <- annotateWithFeatureFlank(as(df_hyper,"GRanges"),
    cgi.obj$CpGi,cgi.obj$shores,
    feature.name="CpGi",flank.name="shores")
  cgiDMRann_hypo <- annotateWithFeatureFlank(as(df_hypo,"GRanges"),
    cgi.obj$CpGi,cgi.obj$shores,
    feature.name="CpGi",flank.name="shores")

  regDMRann_hyper <- annotateWithFeatures(as(df_hyper,"GRanges"),reg.obj)
  regDMRann_hypo <- annotateWithFeatures(as(df_hypo,"GRanges"),reg.obj)

  # Define the output file names
  output_file_gene <- paste0(output_dir, "/geneDMRann_piechart_", name, ".png")
  output_file_cgi <- paste0(output_dir, "/cgiDMRann_piechart_", name, ".png")
  output_file_reg <- paste0(output_dir, "/regDMRann_piechart_", name, ".png")

  # Specify width and height in pixels
  width <- 800
  height <- 600

  # Plot for both hypermethylated and hypomethylated regions
  png(output_file_gene, width = width, height = height)
  par(mfrow = c(1, 2))
  genomation::plotTargetAnnotation(geneDMRann_hyper, main = paste("Hypermethylated (R vs S) -", percent_hyper, "%"))
  genomation::plotTargetAnnotation(geneDMRann_hypo, main = paste("Hypomethylated (R vs S) -", percent_hypo, "%"))
  dev.off()

  png(output_file_cgi, width = width, height = height)
  par(mfrow = c(1, 2))
  genomation::plotTargetAnnotation(cgiDMRann_hyper, main = paste("Hypermethylated (R vs S) -", percent_hyper, "%"))
  genomation::plotTargetAnnotation(cgiDMRann_hypo, main = paste("Hypomethylated (R vs S) -", percent_hypo, "%"))
  dev.off()

  # Define colors for each type of regulatory feature
  feature_colors = c("promoter-like signature" = "red", "proximal enhancer-like signature" = "orange", "distal enhancer-like signature" = "yellow", "DNase-H3K4me3" = "pink", "CTCF-only" = "blue")

  # Plot target annotation with specified colors
  png(output_file_reg, width = width, height = height)
  par(mfrow = c(1, 2))
  genomation::plotTargetAnnotation(regDMRann_hyper, main = paste("Hypermethylated (R vs S) -", percent_hyper, "%"))
  genomation::plotTargetAnnotation(regDMRann_hypo, main = paste("Hypomethylated (R vs S) -", percent_hypo, "%"))
  dev.off()
}
```

```{r}
# R
enrich_regions <- function(df, output_dir, name, dge_df) {
  res <- great(as(df,"GRanges"), "MSigDB:H", "txdb:hg38")
  
  # Define the output file names
  volcano_plot_file <- paste0(output_dir, "/volcano_plot_", name, ".png")
  region_gene_associations_plot_file <- paste0(output_dir, "/region_gene_associations_plot_", name, ".png")
  enrichment_table_file <- paste0(output_dir, "/enrichment_table_", name, ".csv")
  region_gene_associations_file <- paste0(output_dir, "/region_gene_associations_", name, ".csv")
  
  # Save the volcano plot
  png(volcano_plot_file)
  plotVolcano(res)
  dev.off()
  
  # Save the region-gene associations plot
  png(region_gene_associations_plot_file)
  plotRegionGeneAssociations(res)
  dev.off()
  
  # Save the enrichment table
  tb <- getEnrichmentTable(res)
  write.csv(tb, file = enrichment_table_file)
  
  # Save the region-gene associations
  rg <- getRegionGeneAssociations(res)
  # Convert the GRanges object to a data frame
  rg_df <- as.data.frame(rg)

  # Expand the DataFrame
  rg_df <- tidyr::unnest(rg_df, annotated_genes, dist_to_TSS)

  rg_df <- merge(rg_df, dge_df, by.x = "annotated_genes", by.y = "name", all.x = TRUE)
  
  # Write the data frame to a file
  write.csv(rg_df, file = region_gene_associations_file)

  return(rg_df)
}
```

```{r}
link_meth_to_dge <- function(merged_df, output_dir, name) {

  # Remove rows with NA values
  merged_df <- na.omit(merged_df)

  # Define the bins
  bins <- c(-Inf, -500000, -50000, -5000, 0, 5000, 50000, 500000, Inf)

  # Define the labels
  labels <- c("<-500 (kb)", "-500 to -50 (kb)", "-50 to -5 (kb)", "-5 to 0 (kb)", "0 to 5 (kb)", "5 to 50 (kb)", "50 to 500 (kb)", ">500 (kb)")
  # Add a new column "dist_to_promoter_bin" to the dataframe
  merged_df$dist_to_promoter_bin <- cut(merged_df$dist_to_TSS, bins, include.lowest = TRUE, labels = labels)

  new_labels <- c("<-500 (kb)", "-500 to -50 (kb)", "-50 to -5 (kb)", "-5 to 0 (kb)", "0 (kb)", "0 to 5 (kb)", "5 to 50 (kb)", "50 to 500 (kb)", ">500 (kb)")
  
  # Add "0" as a level to the factor
  merged_df$dist_to_promoter_bin <- factor(merged_df$dist_to_promoter_bin, levels = new_labels)

  # Add a group for zero
  merged_df$dist_to_promoter_bin[merged_df$dist_to_TSS == 0] <- "0 (kb)"

  link_plot_file <- paste0(output_dir, "/link_plot_", name, ".png")

  filter_and_plot <- function(df, output_file) {
    # Filter for each combination of methylation and dge status
    hypo_down_df <- df[df$direction == "U" & df$log2FoldChange < 0, ]
    hypo_up_df <- df[df$direction == "U" & df$log2FoldChange > 0, ]
    hyper_down_df <- df[df$direction == "M" & df$log2FoldChange < 0, ]
    hyper_up_df <- df[df$direction == "M" & df$log2FoldChange > 0, ]

    # Calculate the fractions
    hypo_down_frac_upstream <- sum(hypo_down_df$dist_to_TSS <= 0) / nrow(hypo_down_df)
    hypo_down_frac_downstream <- 1 - hypo_down_frac_upstream
    hypo_up_frac_upstream <- sum(hypo_up_df$dist_to_TSS <= 0) / nrow(hypo_up_df)
    hypo_up_frac_downstream <- 1 - hypo_up_frac_upstream
    hyper_down_frac_upstream <- sum(hyper_down_df$dist_to_TSS <= 0) / nrow(hyper_down_df)
    hyper_down_frac_downstream <- 1 - hyper_down_frac_upstream
    hyper_up_frac_upstream <- sum(hyper_up_df$dist_to_TSS <= 0) / nrow(hyper_up_df)
    hyper_up_frac_downstream <- 1 - hyper_up_frac_upstream

    # Create a new variable that categorizes the dist_to_TSS values
    hypo_down_df$Region <- ifelse(hypo_down_df$dist_to_TSS > -50000 & hypo_down_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hypo_down_df$dist_to_TSS > 0 & hypo_down_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))
    hypo_up_df$Region <- ifelse(hypo_up_df$dist_to_TSS > -50000 & hypo_up_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hypo_up_df$dist_to_TSS > 0 & hypo_up_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))
    hyper_down_df$Region <- ifelse(hyper_down_df$dist_to_TSS > -50000 & hyper_down_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hyper_down_df$dist_to_TSS > 0 & hyper_down_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))
    hyper_up_df$Region <- ifelse(hyper_up_df$dist_to_TSS > -50000 & hyper_up_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hyper_up_df$dist_to_TSS > 0 & hyper_up_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))

    values = c("Short-Range Associations (Promoter)" = "black", "Short-Range Associations (Gene Body)" = "red", "Distal associations" = "grey")

    # Create a bar plot for each filtered data
    plot_hypo_down <- ggplot(hypo_down_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hypo-Downregulated\nUpstream + TSS: ", round(hypo_down_frac_upstream, 2), "\nDownstream: ", round(hypo_down_frac_downstream, 2)), x = "Distance from TSS", y = "Count")

    plot_hypo_up <- ggplot(hypo_up_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hypo-Upregulated\nUpstream + TSS: ", round(hypo_up_frac_upstream, 2), "\nDownstream: ", round(hypo_up_frac_downstream, 2)), x = "Distance from TSS", y = "Count")

    plot_hyper_down <- ggplot(hyper_down_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hyper-Downregulated\nUpstream + TSS: ", round(hyper_down_frac_upstream, 2), "\nDownstream: ", round(hyper_down_frac_downstream, 2)), x = "Distance from TSS", y = "Count")

    plot_hyper_up <- ggplot(hyper_up_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hyper-Upregulated\nUpstream + TSS: ", round(hyper_up_frac_upstream, 2), "\nDownstream: ", round(hyper_up_frac_downstream, 2)), x = "Distance from TSS", y = "Count")
    
    # Save the plots
    png(output_file, width = 800, height = 600)
    grid.arrange(plot_hypo_up, plot_hypo_down, plot_hyper_down, plot_hyper_up, ncol = 2)
    dev.off()
  }

  # Use the nested function to filter and plot the data for distances <= 500
  filter_and_plot(merged_df, link_plot_file)
}
```

```{r}
create_karyogram <- function(gr, output_dir, name) {
  # Add a new column for the color
  gr$color <- ifelse(gr$direction == "M", "red", "blue") 

  gr <- as(gr, "GRanges")

  # Load the cytoband data for the hg38 assembly
  hg38_cyto <- getChromInfoFromUCSC("hg38", as.Seqinfo = TRUE, assembled.molecules.only = TRUE)

  output_file <- paste0(output_dir, "/karyogram_", name, ".png")

  # Open a PNG device
  png(output_file)

  # Create a karyogram
  p <- autoplot(hg38_cyto, layout = "karyogram") + layout_karyogram(gr, col = gr$color)
  print(p)

  # Close the PNG device
  dev.off()
}
```

## Roberto

```{r}
directories <- c("/data/lvisser/wgbs_tools/outputs/dmr/MM1_only",
  "/data/lvisser/wgbs_tools/outputs/dmr/NB1_only", 
  "/data/lvisser/wgbs_tools/outputs/dmr/roberto_only")

# Initialize an empty list
all_bed_data <- list()

# R
for (directory in directories) {
  bed_files <- list.files(directory, pattern = "\\.bed$", recursive = TRUE, full.names = TRUE)
  print(bed_files)  # Print file paths
  dir_bed_data <- NULL
  # R
  for (bed_file in bed_files) {
    if (!file.exists(bed_file)) {
      print(paste("File does not exist:", bed_file))
      next
    }
    if (file.info(bed_file)$size == 0) {
      print(paste("File is empty:", bed_file))
      next
    }
    if (length(readLines(bed_file)) <= 1) {
      print(paste("File only contains header:", bed_file))
      next
    }
    bed_data <- readGeneric(bed_file, header = TRUE, keep.all.metadata = TRUE)
    bed_data_df <- as.data.frame(bed_data)  # Convert to data frame
    dir_bed_data <- rbind(dir_bed_data, bed_data_df)
  }
  # Add the dataframe to the list with the directory name as the key
  all_bed_data[[directory]] <- dir_bed_data
}
```

```{r}
dge_names <- c("R_S_MM", "R_S_NB", "R_Ss_sensSHY")
merged_list <- list()

for (i in 1:length(all_bed_data)) {
  name <- basename(directories[i])
  output_dir_name <- paste0(output_dir, "/", name)
  dir.create(output_dir_name, showWarnings = FALSE)
  annotate_regions(all_bed_data[[i]], output_dir_name, name, gene.obj, cgi.obj, reg.obj)
  merged_df <- enrich_regions(as(all_bed_data[[i]],"GRanges"), output_dir_name, name, DE_results_list[[dge_names[i]]])
  merged_list[[name]] <- merged_df
  link_meth_to_dge(merged_list[[i]], output_dir_name, name)
  create_karyogram(merged_list[[i]], output_dir_name, name)
}
```

## Model Features

```{r}
read_shap_csv <- function(file_path) {
  df <- read.csv(file_path) %>%
    filter(z_score >= 2)

  df <- df %>%
    separate("segment_id", into = c("chromosome", "start", "end"), sep = "[:-]")

  return(df)
}
```

```{r}
files <- c("/data/lvisser/shap/shap_wgbs_mm1_df_annot.csv",
  "/data/lvisser/shap/shap_wgbs_nb1_df_annot.csv", 
  "/data/lvisser/shap/shap_wgbs_rob_df_annot.csv")

# Initialize an empty list
all_shap_data <- list()

for (file in files) {
  shap_data <- read_shap_csv(file)
  all_shap_data[[file]] <- shap_data
}
```

```{r}
dge_names <- c("R_S_MM", "R_S_NB", "R_Ss_sensSHY")
merged_list <- list()

for (i in 1:length(all_shap_data)) {
  name <- sub("\\.csv$", "", basename(files[i]))
  output_dir_name <- paste0(output_dir, "/", name)
  dir.create(output_dir_name, showWarnings = FALSE)
  annotate_regions(all_shap_data[[i]], output_dir_name, name, gene.obj, cgi.obj, reg.obj)
  merged_df <- enrich_regions(as(all_shap_data[[i]],"GRanges"), output_dir_name, name, DE_results_list[[dge_names[i]]])
  merged_list[[name]] <- merged_df
  link_meth_to_dge(merged_list[[i]], output_dir_name, name)
  create_karyogram(merged_list[[i]], output_dir_name, name)
}
```


