
## Import packages
```{r}
library(rGREAT)
library(genomation)
library(methylKit)
library(tidyverse)
library(ggbio)
library(GenomicFeatures)
library(GenomicRanges)
library(ggplot2)
library(gridExtra)

output_dir <- "/data/lvisser/post_analysis"

whitelist <- read.csv("/data/lvisser/ferroptosis_gene_whitelists/complete_gene_list_no_duplicates_cleaned.txt", header = FALSE)
```

```{r}
DE_results_human_Group_R_S_MM <- read.csv("/data/lvisser/DE_genes/DE_results_human_Group_R_S_MM.csv") %>%
  filter(padj < 0.05) %>%
  dplyr::select(baseMean,log2FoldChange,lfcSE,pvalue,padj,name,type,description)
DE_results_human_Group_R_S_NB <- read.csv("/data/lvisser/DE_genes/DE_results_human_Group_R_S_NB.csv") %>%
  filter(padj < 0.05) %>%
  dplyr::select(baseMean,log2FoldChange,lfcSE,pvalue,padj,name,type,description)
DE_results_human_Group_R_Ss_sensSHY <- read.csv("/data/lvisser/DE_genes/DE_results_human_Group_R_Ss_sensSHY.csv") %>%
  filter(padj < 0.05) %>%
  dplyr::select(baseMean,log2FoldChange,lfcSE,pvalue,padj,name,type,description)

# Create a list of dataframes
DE_results_list <- list(
  "R_S_MM" = DE_results_human_Group_R_S_MM,
  "R_S_NB" = DE_results_human_Group_R_S_NB,
  "R_Ss_sensSHY" = DE_results_human_Group_R_Ss_sensSHY
)
```

```{r}
gene.obj <- readTranscriptFeatures("/data/lvisser/segmentations/genes_refseq_all_ucsc.bed")
head(gene.obj)
```

```{r}
cgi.obj <- readFeatureFlank("/data/lvisser/segmentations/cpg_islands_ucsc_cleaned.bed",
                           feature.flank.name=c("CpGi","shores"))
head(cgi.obj)
```

```{r}
reg.obj <- readGeneric("/data/lvisser/segmentations/encode_ccres_ucsc.bed", keep.all.metadata = TRUE, header = TRUE)
reg.obj <- split(reg.obj, mcols(reg.obj)$ucscLabel)
head(reg.obj)
```

```{r}
annotate_regions <- function(df, whitelist_df, output_dir, name, gene.obj, cgi.obj, reg.obj) {
  # Separate hypermethylated and hypomethylated regions
  df_hyper <- df[df$direction == "M",]
  df_hypo <- df[df$direction == "U",]

  # Keep only one row for each segment_id
  whitelist_df <- dplyr::distinct(whitelist_df, segment_id, .keep_all = TRUE)

  # Get the unique segment_ids in df and whitelist_df
  unique_segment_ids_df <- unique(df$segment_id)
  unique_segment_ids_whitelist_df <- unique(whitelist_df$segment_id)

  # Get the number of unique segment_ids in df that are also in whitelist_df
  num_overlap <- sum(unique_segment_ids_df %in% unique_segment_ids_whitelist_df)

  # Calculate the percentage
  percent_overlap <- round(num_overlap / length(unique_segment_ids_df) * 100, 2)
  
  whitelist_df_hyper <- whitelist_df[whitelist_df$direction == "M",]
  whitelist_df_hypo <- whitelist_df[whitelist_df$direction == "U",]

  # Calculate percentages
  percent_hyper <- round(nrow(df_hyper) / nrow(df) * 100, 2)
  percent_hypo <- round(nrow(df_hypo) / nrow(df) * 100, 2)

  percent_hyper_whitelist <- round(nrow(whitelist_df_hyper) / nrow(whitelist_df) * 100, 2)
  percent_hypo_whitelist <- round(nrow(whitelist_df_hypo) / nrow(whitelist_df) * 100, 2)

  # Annotate regions for both hypermethylated and hypomethylated
  geneDMRann_hyper <- annotateWithGeneParts(as(df_hyper,"GRanges"),gene.obj)
  geneDMRann_hypo <- annotateWithGeneParts(as(df_hypo,"GRanges"),gene.obj)

  cgiDMRann_hyper <- annotateWithFeatureFlank(as(df_hyper,"GRanges"),
    cgi.obj$CpGi,cgi.obj$shores,
    feature.name="CpGi",flank.name="shores")
  cgiDMRann_hypo <- annotateWithFeatureFlank(as(df_hypo,"GRanges"),
    cgi.obj$CpGi,cgi.obj$shores,
    feature.name="CpGi",flank.name="shores")

  addOtherCategory <- function(annotated_obj) {
    annotated_obj@annotation["other"] <- 100 - sum(annotated_obj@precedence)
    annotated_obj@precedence["other"] <- 100 - sum(annotated_obj@precedence)
    annotated_obj@num.annotation["other"] <- Inf
    annotated_obj@num.precedence["other"] <- Inf
    return(annotated_obj)
  }

  regDMRann_hyper <- annotateWithFeatures(as(df_hyper,"GRanges"),reg.obj)
  regDMRann_hypo <- annotateWithFeatures(as(df_hypo,"GRanges"),reg.obj)
  regDMRann_hyper <- addOtherCategory(regDMRann_hyper)
  regDMRann_hypo <- addOtherCategory(regDMRann_hypo)

  # Define the minimum number of rows
  min_rows <- 3

  # Filter the data frames
  if (nrow(whitelist_df_hyper) >= min_rows) {
    geneDMRann_hyper_whitelist <- annotateWithGeneParts(as(whitelist_df_hyper,"GRanges"), gene.obj)
    
    cgiDMRann_hyper_whitelist <- annotateWithFeatureFlank(as(whitelist_df_hyper,"GRanges"), cgi.obj$CpGi,cgi.obj$shores, feature.name="CpGi",flank.name="shores")
    
    regDMRann_hyper_whitelist <- annotateWithFeatures(as(whitelist_df_hyper,"GRanges"),reg.obj)
  
    regDMRann_hyper_whitelist <- addOtherCategory(regDMRann_hyper_whitelist)
  }

  # Only run hypo if whitelist_df_hypo has enough rows
  if (nrow(whitelist_df_hypo) >= min_rows) {
    geneDMRann_hypo_whitelist <- annotateWithGeneParts(as(whitelist_df_hypo,"GRanges"), gene.obj)

    cgiDMRann_hypo_whitelist <- annotateWithFeatureFlank(as(whitelist_df_hypo,"GRanges"), cgi.obj$CpGi,cgi.obj$shores, feature.name="CpGi",flank.name="shores")

    regDMRann_hypo_whitelist <- annotateWithFeatures(as(whitelist_df_hypo,"GRanges"),reg.obj)

    regDMRann_hypo_whitelist <- addOtherCategory(regDMRann_hypo_whitelist)
  }

  # Define the output file names
  output_file_gene <- paste0(output_dir, "/geneDMRann_piechart_", name, ".png")
  output_file_cgi <- paste0(output_dir, "/cgiDMRann_piechart_", name, ".png")
  output_file_reg <- paste0(output_dir, "/regDMRann_piechart_", name, ".png")

  # Specify width and height in pixels
  width <- 800
  height <- 600

  # Adjust the margins of the plots
  par(mar = c(1, 1, 1, 1), oma = c(0, 0, 0, 0))

  # Add the subplots to the existing plots
  png(output_file_gene, width = width, height = height)
  par(mfrow = c(2, 2))
  genomation::plotTargetAnnotation(geneDMRann_hyper, main = paste("Hypermethylated (R vs S) - All -", percent_hyper, "%"))
  genomation::plotTargetAnnotation(geneDMRann_hypo, main = paste("Hypomethylated (R vs S) - All - ", percent_hypo, "%"))
  if (nrow(whitelist_df_hyper) >= min_rows) {
    genomation::plotTargetAnnotation(geneDMRann_hyper_whitelist, main = paste("Hypermethylated (R vs S) - FRGs -", percent_hyper_whitelist, "%"))
  }
  if (nrow(whitelist_df_hypo) >= min_rows) {
    genomation::plotTargetAnnotation(geneDMRann_hypo_whitelist, main = paste("Hypomethylated (R vs S) - FRGs -", percent_hypo_whitelist, "%"))
  }
  mtext(paste("Percentage of all segments annotated with at least one FRG: ", percent_overlap, "%"), side = 1, line = -2, outer = TRUE)
  dev.off()

  # Add the subplots to the existing plots
  png(output_file_cgi, width = width, height = height)
  par(mfrow = c(2, 2))
  genomation::plotTargetAnnotation(cgiDMRann_hyper, main = paste("Hypermethylated (R vs S) - All -", percent_hyper, "%"))
  genomation::plotTargetAnnotation(cgiDMRann_hypo, main = paste("Hypomethylated (R vs S) - All -", percent_hypo, "%"))
  if (nrow(whitelist_df_hyper) >= min_rows) {
    genomation::plotTargetAnnotation(cgiDMRann_hyper_whitelist, main = paste("Hypermethylated (R vs S) - FRGs -", percent_hyper_whitelist, "%"))
  }
  if (nrow(whitelist_df_hypo) >= min_rows) {
    genomation::plotTargetAnnotation(cgiDMRann_hypo_whitelist, main = paste("Hypomethylated (R vs S) - FRGs -", percent_hypo_whitelist, "%"))
  }
  mtext(paste("Percentage of all segments annotated with at least one FRG: ", percent_overlap, "%"), side = 1, line = -2, outer = TRUE)
  dev.off()

  png(output_file_reg, width = width, height = height)
  par(mfrow = c(2, 2))
  genomation::plotTargetAnnotation(regDMRann_hyper, main = paste("Hypermethylated (R vs S) - All -", percent_hyper, "%"))
  genomation::plotTargetAnnotation(regDMRann_hypo, main = paste("Hypomethylated (R vs S) - All -", percent_hypo, "%"))
  if (nrow(whitelist_df_hyper) >= min_rows) {
    genomation::plotTargetAnnotation(regDMRann_hyper_whitelist, main = paste("Hypermethylated (R vs S) - FRGs -", percent_hyper_whitelist, "%"))
  }
  if (nrow(whitelist_df_hypo) >= min_rows) {
    genomation::plotTargetAnnotation(regDMRann_hypo_whitelist, main = paste("Hypomethylated (R vs S) - FRGs -", percent_hypo_whitelist, "%"))
  }
  mtext(paste("Percentage of all segments annotated with at least one FRG: ", percent_overlap, "%"), side = 1, line = -2, outer = TRUE)
  dev.off()

  annotations_to_df <- function(segment_ids, obj_list) {
    # Initialize an empty dataframe with segment_ids
    df_annotations <- data.frame(segment_id = segment_ids)
    
    # Iterate over each members dataframe in the list
    for (name in names(obj_list)) {
      obj <- obj_list[[name]]

      members <- obj@members

      if (is.list(members)) {
        # Access the matrix in the S4 object
        members <- as.data.frame(obj@members)
      }
      
      # Get the column names where each row is 1
      annotations <- apply(members, 1, function(x) {
        names(x)[which(x == 1)]
      })
      
      # Combine the annotations into a comma-separated string
      annotations <- sapply(annotations, function(x) {
        paste(x, collapse = ",")
      })
      
      # Add the annotations as a new column in the dataframe
      df_annotations[[paste(name, "members", sep = "_")]] <- annotations

      if (name == "gene") {
        dist_to_tss <- obj@dist.to.TSS[, c("dist.to.feature", "feature.name")]
        df_annotations[[paste(name, "dist.to.feature", sep = "_")]] <- dist_to_tss[, "dist.to.feature"]
        df_annotations[[paste(name, "feature.name", sep = "_")]] <- dist_to_tss[, "feature.name"]
      }
    }
    
    return(df_annotations)
  }
  hyper_obj_list = list(
    gene = geneDMRann_hyper, 
    cgi = cgiDMRann_hyper, 
    reg = regDMRann_hyper
  )

  hypo_obj_list = list(
    gene = geneDMRann_hypo, 
    cgi = cgiDMRann_hypo, 
    reg = regDMRann_hypo
  )

  hyper_df_annot = annotations_to_df(df_hyper$segment_id, hyper_obj_list)
  hypo_df_annot = annotations_to_df(df_hypo$segment_id, hypo_obj_list)

  df_combined <- rbind(hyper_df_annot, hypo_df_annot)

  # Left join df with hyper_df_annot
  df <- merge(df, df_combined, by = "segment_id", all.x = TRUE)

  df <- df[order(df$z_score, decreasing = TRUE), ] %>%
    distinct()

  region_gene_associations_file <- paste0(output_dir, "/region_gene_associations_", name, ".csv")

  write.csv(df, file = region_gene_associations_file)

  hyper_whitelist_obj_list = list(
    gene = geneDMRann_hyper_whitelist, 
    cgi = cgiDMRann_hyper_whitelist, 
    reg = regDMRann_hyper_whitelist
  )

  hypo_whitelist_obj_list = list(
    gene = geneDMRann_hypo_whitelist, 
    cgi = cgiDMRann_hypo_whitelist, 
    reg = regDMRann_hypo_whitelist
  )

  hyper_whitelist_df_annot = annotations_to_df(whitelist_df_hyper$segment_id, hyper_whitelist_obj_list)
  hypo_whitelist_df_annot = annotations_to_df(whitelist_df_hypo$segment_id, hypo_whitelist_obj_list)

  whitelist_df_combined <- rbind(hyper_whitelist_df_annot, hypo_whitelist_df_annot)

  # Left join df with hyper_df_annot
  whitelist_df <- merge(whitelist_df, whitelist_df_combined, by = "segment_id", all.x = TRUE)

  whitelist_df <- whitelist_df[order(whitelist_df$z_score, decreasing = TRUE), ] %>%
    distinct()

  write.csv(whitelist_df, file = paste0(output_dir, "/region_gene_associations_whitelist_", name, ".csv"))
}
```

```{r}
# R
enrich_regions <- function(df, output_dir, name, dge_df = NULL, whitelist) {
  res <- great(as(df,"GRanges"), "MSigDB:H", "txdb:hg38")
  
  # Define the output file names
  volcano_plot_file <- paste0(output_dir, "/volcano_plot_", name, ".png")
  region_gene_associations_plot_file <- paste0(output_dir, "/region_gene_associations_plot_", name, ".png")
  enrichment_table_file <- paste0(output_dir, "/enrichment_table_", name, ".csv")
  region_gene_associations_file <- paste0(output_dir, "/region_gene_associations_", name, ".csv")
  
  # Save the volcano plot
  png(volcano_plot_file)
  plotVolcano(res)
  dev.off()
  
  # Save the region-gene associations plot
  png(region_gene_associations_plot_file)
  plotRegionGeneAssociations(res)
  dev.off()
  
  # Save the enrichment table
  tb <- getEnrichmentTable(res)
  write.csv(tb, file = enrichment_table_file)
  
  # Save the region-gene associations
  rg <- getRegionGeneAssociations(res)
  # Convert the GRanges object to a data frame
  rg_df <- as.data.frame(rg)

  # Expand the DataFrame
  rg_df <- tidyr::unnest(rg_df, annotated_genes, dist_to_TSS)

  if (!is.null(dge_df)) {
    rg_df <- merge(rg_df, dge_df, by.x = "annotated_genes", by.y = "name", all.x = TRUE)
  }

  # Filter rg_df based on the whitelist
  rg_df_whitelist <- rg_df[rg_df$annotated_genes %in% whitelist[, 1],]

  # Write the data frame to a file
  write.csv(rg_df, file = region_gene_associations_file)

  write.csv(rg_df_whitelist, file = paste0(output_dir, "/region_gene_associations_whitelist_", name, ".csv"))

  return(list(rg_df = rg_df, rg_df_whitelist = rg_df_whitelist))
}
```

```{r}
link_meth_to_dge <- function(merged_df, output_dir, name) {

  # Remove rows with NA values
  merged_df <- na.omit(merged_df)

  # Define the bins
  bins <- c(-Inf, -500000, -50000, -5000, 0, 5000, 50000, 500000, Inf)

  # Define the labels
  labels <- c("<-500 (kb)", "-500 to -50 (kb)", "-50 to -5 (kb)", "-5 to 0 (kb)", "0 to 5 (kb)", "5 to 50 (kb)", "50 to 500 (kb)", ">500 (kb)")
  # Add a new column "dist_to_promoter_bin" to the dataframe
  merged_df$dist_to_promoter_bin <- cut(merged_df$dist_to_TSS, bins, include.lowest = TRUE, labels = labels)

  new_labels <- c("<-500 (kb)", "-500 to -50 (kb)", "-50 to -5 (kb)", "-5 to 0 (kb)", "0 (kb)", "0 to 5 (kb)", "5 to 50 (kb)", "50 to 500 (kb)", ">500 (kb)")
  
  # Add "0" as a level to the factor
  merged_df$dist_to_promoter_bin <- factor(merged_df$dist_to_promoter_bin, levels = new_labels)

  # Add a group for zero
  merged_df$dist_to_promoter_bin[merged_df$dist_to_TSS == 0] <- "0 (kb)"

  link_plot_file <- paste0(output_dir, "/link_plot_", name, ".png")

  filter_and_plot <- function(df, output_file) {
    # Filter for each combination of methylation and dge status
    hypo_down_df <- df[df$direction == "U" & df$log2FoldChange < 0, ]
    hypo_up_df <- df[df$direction == "U" & df$log2FoldChange > 0, ]
    hyper_down_df <- df[df$direction == "M" & df$log2FoldChange < 0, ]
    hyper_up_df <- df[df$direction == "M" & df$log2FoldChange > 0, ]

    # Calculate the fractions
    hypo_down_frac_upstream <- sum(hypo_down_df$dist_to_TSS <= 0) / nrow(hypo_down_df)
    hypo_down_frac_downstream <- 1 - hypo_down_frac_upstream
    hypo_up_frac_upstream <- sum(hypo_up_df$dist_to_TSS <= 0) / nrow(hypo_up_df)
    hypo_up_frac_downstream <- 1 - hypo_up_frac_upstream
    hyper_down_frac_upstream <- sum(hyper_down_df$dist_to_TSS <= 0) / nrow(hyper_down_df)
    hyper_down_frac_downstream <- 1 - hyper_down_frac_upstream
    hyper_up_frac_upstream <- sum(hyper_up_df$dist_to_TSS <= 0) / nrow(hyper_up_df)
    hyper_up_frac_downstream <- 1 - hyper_up_frac_upstream

    # Create a new variable that categorizes the dist_to_TSS values
    hypo_down_df$Region <- ifelse(hypo_down_df$dist_to_TSS > -50000 & hypo_down_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hypo_down_df$dist_to_TSS > 0 & hypo_down_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))
    hypo_up_df$Region <- ifelse(hypo_up_df$dist_to_TSS > -50000 & hypo_up_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hypo_up_df$dist_to_TSS > 0 & hypo_up_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))
    hyper_down_df$Region <- ifelse(hyper_down_df$dist_to_TSS > -50000 & hyper_down_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hyper_down_df$dist_to_TSS > 0 & hyper_down_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))
    hyper_up_df$Region <- ifelse(hyper_up_df$dist_to_TSS > -50000 & hyper_up_df$dist_to_TSS <= 0, "Short-Range Associations (Promoter)", ifelse(hyper_up_df$dist_to_TSS > 0 & hyper_up_df$dist_to_TSS <= 50000, "Short-Range Associations (Gene Body)", "Distal associations"))

    values = c("Short-Range Associations (Promoter)" = "black", "Short-Range Associations (Gene Body)" = "red", "Distal associations" = "grey")

    # Create a bar plot for each filtered data
    plot_hypo_down <- ggplot(hypo_down_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hypo-Downregulated\nUpstream + TSS: ", round(hypo_down_frac_upstream, 2), "\nDownstream: ", round(hypo_down_frac_downstream, 2)), x = "Distance from TSS", y = "Count")

    plot_hypo_up <- ggplot(hypo_up_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hypo-Upregulated\nUpstream + TSS: ", round(hypo_up_frac_upstream, 2), "\nDownstream: ", round(hypo_up_frac_downstream, 2)), x = "Distance from TSS", y = "Count")

    plot_hyper_down <- ggplot(hyper_down_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hyper-Downregulated\nUpstream + TSS: ", round(hyper_down_frac_upstream, 2), "\nDownstream: ", round(hyper_down_frac_downstream, 2)), x = "Distance from TSS", y = "Count")

    plot_hyper_up <- ggplot(hyper_up_df, aes(x = dist_to_promoter_bin, fill = Region)) + 
      geom_bar(stat = "count") + 
      scale_x_discrete(drop = FALSE) +
      scale_fill_manual(values = values) +
      theme(axis.text.x = element_text(angle = 90)) +
      labs(title = paste0("Hyper-Upregulated\nUpstream + TSS: ", round(hyper_up_frac_upstream, 2), "\nDownstream: ", round(hyper_up_frac_downstream, 2)), x = "Distance from TSS", y = "Count")
  
    png(output_file, width = 800, height = 600)
    grid.arrange(plot_hypo_up, plot_hypo_down, plot_hyper_down, plot_hyper_up, ncol = 2)
    dev.off()
  }

  # Use the nested function to filter and plot the data for distances <= 500
  filter_and_plot(merged_df, link_plot_file)
}
```

```{r}
create_karyogram <- function(gr, output_dir, name, weight_col = NULL) {
  # Add a new column for the color
  gr$color <- ifelse(gr$direction == "M", "red", "blue") 

  # Add a new column for the alpha level based on the weight, if provided
  if (!is.null(weight_col)) {
    # Scale the weight column to the range [0,1]
    gr$alpha <- scales::rescale(gr[[weight_col]], to = c(0, 1))
    # Convert the color and alpha level to a single color specification
    gr$color <- ifelse(gr$direction == "M", rgb(1, 0, 0, gr$alpha), rgb(0, 0, 1, gr$alpha))
  }

  gr <- as(gr, "GRanges")

  # Load the cytoband data for the hg38 assembly
  hg38_cyto <- getChromInfoFromUCSC("hg38", as.Seqinfo = TRUE, assembled.molecules.only = TRUE)

  output_file <- paste0(output_dir, "/karyogram_", name, ".png")

  # Open a PNG device
  png(output_file)

  # Create a karyogram
  p <- autoplot(hg38_cyto, layout = "karyogram") + layout_karyogram(gr, col = gr$color)
  print(p)

  # Close the PNG device
  dev.off()
}
```

## Segments

```{r}
directories <- c("/data/lvisser/wgbs_tools/outputs/segmentations/MM_only",
  "/data/lvisser/wgbs_tools/outputs/segmentations/NB_only", 
  "/data/lvisser/wgbs_tools/outputs/segmentations/roberto_only",
  "/data/lvisser/wgbs_tools/outputs/segmentations/NB_and_MM",
  "/data/lvisser/wgbs_tools/outputs/segmentations/NB_and_roberto")

# Initialize an empty list
all_seg_data <- list()

# R
for (directory in directories) {
  seg_files <- list.files(directory, pattern = "\\.tsv$", recursive = TRUE, full.names = TRUE)
  print(seg_files)  # Print file paths
  dir_seg_data <- NULL
  # R
  for (seg_file in seg_files) {
    if (!file.exists(seg_file)) {
      print(paste("File does not exist:", seg_file))
      next
    }
    if (file.info(seg_file)$size == 0) {
      print(paste("File is empty:", seg_file))
      next
    }
    if (length(readLines(seg_file)) <= 1) {
      print(paste("File only contains header:", seg_file))
      next
    }
    seg_data <- readGeneric(seg_file, header = TRUE, keep.all.metadata = TRUE)
    seg_data_df <- as.data.frame(seg_data)  # Convert to data frame
    dir_seg_data <- rbind(dir_seg_data, seg_data_df)
  }
  # Add the dataframe to the list with the directory name as the key
  # Remove rows with NA values
  dir_seg_data <- na.omit(dir_seg_data)
  all_seg_data[[directory]] <- dir_seg_data
}
```

```{r}
for (i in 1:length(all_seg_data)) {
  name <- paste0(basename(directories[i]), "_seg")
  output_dir_name <- paste0(output_dir, "/", name)
  dir.create(output_dir_name, showWarnings = FALSE)
  merged_df <- enrich_regions(as(all_seg_data[[i]],"GRanges"), output_dir_name, name)
}
```

```{r}
output_dir_name <- "/data/lvisser/post_analysis/rwgbs_rob_seg"
merged_df <- enrich_regions(as(all_seg_data[[i]],"GRanges"), output_dir_name, name)
```

## DMR

```{r}
directories <- c("/data/lvisser/wgbs_tools/outputs/dmr/MM_only",
  "/data/lvisser/wgbs_tools/outputs/dmr/NB_only", 
  "/data/lvisser/wgbs_tools/outputs/dmr/roberto_only")

# Initialize an empty list
all_bed_data <- list()

# R
for (directory in directories) {
  bed_files <- list.files(directory, pattern = "\\.bed$", recursive = TRUE, full.names = TRUE)
  print(bed_files)  # Print file paths
  dir_bed_data <- NULL
  # R
  for (bed_file in bed_files) {
    if (!file.exists(bed_file)) {
      print(paste("File does not exist:", bed_file))
      next
    }
    if (file.info(bed_file)$size == 0) {
      print(paste("File is empty:", bed_file))
      next
    }
    if (length(readLines(bed_file)) <= 1) {
      print(paste("File only contains header:", bed_file))
      next
    }
    bed_data <- readGeneric(bed_file, header = TRUE, keep.all.metadata = TRUE)
    bed_data_df <- as.data.frame(bed_data)  # Convert to data frame
    # Add a segment_id column
    bed_data_df <- mutate(bed_data_df, segment_id = paste0(seqnames, ":", start, "-", end))

    dir_bed_data <- rbind(dir_bed_data, bed_data_df)
  }
  # Add the dataframe to the list with the directory name as the key
  all_bed_data[[directory]] <- dir_bed_data
}
```

```{r}
dge_names <- c("R_S_MM", "R_S_NB", "R_Ss_sensSHY")
merged_list <- list()

for (i in 1:length(all_bed_data)) {
  name <- paste0(basename(directories[i]), "_dmr")
  output_dir_name <- paste0(output_dir, "/", name)
  dir.create(output_dir_name, showWarnings = FALSE)
  merged_df_list <- enrich_regions(as(all_bed_data[[i]],"GRanges"), output_dir_name, name, DE_results_list[[dge_names[i]]], whitelist)
  merged_list[[name]] <- merged_df_list[[1]]
  annotate_regions(all_bed_data[[i]], merged_df_list[[2]], output_dir_name, name, gene.obj, cgi.obj, reg.obj)
  link_meth_to_dge(merged_list[[i]], output_dir_name, name)
  create_karyogram(merged_list[[i]], output_dir_name, name)
}
```

## Model Features

```{r}
read_shap_csv <- function(file_path) {
  df <- read.csv(file_path)

  df <- df %>%
    separate("segment_id", into = c("chromosome", "start", "end"), sep = "[:-]", remove = FALSE) %>%
    filter(z_score > 0)

  return(df)
}
```

```{r}
files <- c("/data/lvisser/shap/shap_wgbs_mm_df_annot.csv",
  "/data/lvisser/shap/shap_wgbs_nb_df_annot.csv", 
  "/data/lvisser/shap/shap_wgbs_rob_df_annot.csv")

# Initialize an empty list
all_shap_data <- list()

for (file in files) {
  shap_data <- read_shap_csv(file)
  all_shap_data[[file]] <- shap_data
}
```

```{r}
dge_names <- c("R_S_MM", "R_S_NB", "R_Ss_sensSHY")
merged_list <- list()

for (i in 1:length(all_shap_data)) {
  name <- sub("\\.csv$", "", basename(files[i]))
  output_dir_name <- paste0(output_dir, "/", name)
  dir.create(output_dir_name, showWarnings = FALSE)
  merged_df_list <- enrich_regions(as(all_shap_data[[i]],"GRanges"), output_dir_name, name, DE_results_list[[dge_names[i]]], whitelist)
  merged_list[[name]] <- merged_df_list[[1]]
  annotate_regions(merged_df_list[[1]], merged_df_list[[2]], output_dir_name, name, gene.obj, cgi.obj, reg.obj)
  link_meth_to_dge(merged_list[[i]], output_dir_name, name)
  create_karyogram(merged_list[[i]], output_dir_name, name, "z_score")
}
```